File: api/services/widget_inserter.py
"""
Widget inserter — injects PriceRunner JS widgets and product images into
article markdown at generator-specified anchor positions.

Runs at publish time and in preview — NOT as a pipeline step.
Widget HTML is derived from live brief data; if the widget format changes,
re-publishing regenerates it without re-running Claude.

Widget selection (archive/docs/pricerunner-widgets.md):
  single-product-review → product.js (3 stores, in-stock, national)
  all other article types → singleproduct.js (lowest price, compact)

Placement format from generator JSON output:
  { "type": "image"|"widget", "productId": "<id>",
    "anchor": { "kind": "after-heading"|"end-of-section"|"after-intro"|"before-heading",
                "section": "<H2 text>" } }
"""

import re
import uuid
from urllib.parse import quote

from services.brief_builder import ContentBrief, ImageRef, ProductBrief, get_site_config


def _widget_id() -> str:
    return uuid.uuid4().hex[:8]


def _numeric_id(product_id: str) -> str:
    """Strip 'pr_' prefix to get the raw numeric PriceRunner product ID."""
    return product_id.removeprefix("pr_")


def _slugify(text: str) -> str:
    for src, dst in [("æ", "ae"), ("ø", "oe"), ("å", "aa")]:
        text = text.replace(src, dst)
    text = text.lower()
    text = re.sub(r"[^\w\s-]", "", text)
    text = re.sub(r"[\s_]+", "-", text)
    return text.strip("-")


def render_widget(product: ProductBrief, partner_id: str, article_type: str) -> str:
    """
    PriceRunner JS embed for one product.
    single-product-review → product.js (3 stores); everything else → singleproduct.js.
    """
    pid = _numeric_id(product.id)
    wid = _widget_id()
    encoded_partner = quote(partner_id)

    if article_type == "single-product-review":
        elem_id = f"pr-product-widget-{wid}"
        src = (
            f"https://api.pricerunner.com/publisher-widgets/dk/product.js"
            f"?onlyInStock=true&offerOrigin=NATIONAL&offerLimit=3"
            f"&productId={pid}&partnerId={encoded_partner}&widgetId={elem_id}"
        )
    else:
        elem_id = f"pr-singleproduct-widget-{wid}"
        src = (
            f"https://api.pricerunner.com/publisher-widgets/dk/singleproduct.js"
            f"?productId={pid}&partnerId={encoded_partner}&widgetId={elem_id}"
        )

    # Attribution href: product URL without refsite so the disclosure link is clean
    disclosure_url = re.sub(r"[?&]refsite=[^&]*", "", product.affiliate_url).rstrip("?&")

    disclosure = (
        '<div style="display:inline-block">'
        f'<a href="{disclosure_url}" rel="nofollow">'
        "<p style=\"font:14px 'Klarna Text',Helvetica,sans-serif;font-style:italic;"
        'color:var(--grayscale100);text-decoration:underline;">'
        'Annonce i samarbejde med <span style="font-weight:bold">PriceRunner</span>'
        "</p></a></div>"
    )

    return (
        f'<div id="{elem_id}" style="display:block;width:100%"></div>\n'
        f'<script type="text/javascript" src="{src}" async></script>\n'
        f"{disclosure}"
    )


def render_image(image: ImageRef) -> str:
    return (
        '<figure style="margin:1.5em 0;text-align:center">'
        f'<img src="{image.url}" alt="{image.alt}" '
        'style="max-width:100%;height:auto;border-radius:4px">'
        f'<figcaption style="font-size:0.85em;color:#666;margin-top:0.4em">'
        f"{image.caption}</figcaption>"
        "</figure>"
    )


def _find_heading(lines: list[str], section: str) -> int | None:
    """Find the H1-H3 line matching section text. Tries exact, then slug, then prefix."""
    slug = _slugify(section)
    for i, line in enumerate(lines):
        m = re.match(r"^#{1,3}\s+(.+)", line)
        if not m:
            continue
        text = m.group(1).strip()
        if text == section:
            return i
        text_slug = _slugify(text)
        if text_slug == slug or text_slug.startswith(slug):
            return i
    return None


def _find_section_end(lines: list[str], heading_idx: int) -> int:
    """Line index to insert at end of section: just before next H1/H2, or EOF."""
    for i in range(heading_idx + 1, len(lines)):
        if re.match(r"^#{1,2}\s+", lines[i]):
            j = i - 1
            while j > heading_idx and not lines[j].strip():
                j -= 1
            return j + 1
    return len(lines)


def _find_after_intro(lines: list[str]) -> int | None:
    """Return insertion index after the first paragraph following the H1."""
    past_h1 = False
    in_para = False
    for i, line in enumerate(lines):
        if re.match(r"^#\s+", line):
            past_h1 = True
            continue
        if not past_h1:
            continue
        if line.strip():
            in_para = True
        elif in_para:
            return i + 1
    return None


def insert_anchored_placements(
    article: str,
    brief: ContentBrief,
    placements: list[dict],
) -> tuple[str, list[str]]:
    """
    Insert widget and image HTML at anchor positions in the article markdown.
    Returns (modified_article, errors). Errors are non-fatal — the article is
    returned with whatever insertions succeeded.
    """
    if not placements:
        return article, []

    site_cfg = get_site_config(brief.site_key)
    partner_id = site_cfg.pricerunner_partner_id
    products_by_id: dict[str, ProductBrief] = {p.id: p for p in brief.products}
    images_by_id: dict[str, ImageRef] = {img.product_id: img for img in brief.images}

    lines = article.split("\n")
    errors: list[str] = []
    # (line_index, html_block) — applied in descending order to preserve indices
    insertions: list[tuple[int, str]] = []

    for placement in placements:
        p_type = placement.get("type", "widget")
        product_id = placement.get("productId", "")
        anchor = placement.get("anchor", {})
        kind = anchor.get("kind", "after-heading")
        section = anchor.get("section", "")

        product = products_by_id.get(product_id) or (brief.products[0] if brief.products else None)
        if not product:
            errors.append(f"No product found for productId='{product_id}'")
            continue

        if p_type == "image":
            image = images_by_id.get(product_id) or (brief.images[0] if brief.images else None)
            if not image:
                errors.append(f"No image found for productId='{product_id}'")
                continue
            html_block = render_image(image)
        else:
            html_block = render_widget(product, partner_id, brief.article_type)

        if kind == "after-intro":
            idx = _find_after_intro(lines)
            if idx is None:
                errors.append("Could not locate intro paragraph for after-intro placement")
                continue
            insertions.append((idx, html_block))
        elif kind in ("after-heading", "end-of-section", "before-heading"):
            if not section:
                errors.append(f"Missing section text for kind='{kind}'")
                continue
            h_idx = _find_heading(lines, section)
            if h_idx is None:
                errors.append(f"Heading not found: '{section}'")
                continue
            if kind == "after-heading":
                insertions.append((h_idx + 1, html_block))
            elif kind == "before-heading":
                insertions.append((h_idx, html_block))
            else:  # end-of-section
                insertions.append((_find_section_end(lines, h_idx), html_block))
        else:
            errors.append(f"Unknown anchor kind: '{kind}'")

    # Apply insertions in reverse order so earlier indices stay valid
    insertions.sort(key=lambda x: x[0], reverse=True)
    for idx, html_block in insertions:
        lines.insert(idx, "")
        lines.insert(idx + 1, html_block)
        lines.insert(idx + 2, "")

    return "\n".join(lines), errors

File: archive/src/services/widget-inserter.ts
import { randomUUID } from "crypto";
import type { ContentBrief, Placement, AnchoredPlacement } from "../types/index.js";
import { SITE_CONFIGS } from "../config/sites.js";

// ─── Heading anchor resolver ──────────────────────────────────────────────────

/** Slugify: lowercase, transliterate Danish, strip punctuation */
function toSlug(s: string): string {
  return s
    .toLowerCase()
    .replace(/æ/g, "ae").replace(/ø/g, "oe").replace(/å/g, "aa")
    .replace(/[^a-z0-9\s]/g, "")
    .trim()
    .replace(/\s+/g, "-");
}

/**
 * Parse all headings from a Markdown article.
 * Returns array of { text, lineIndex } where lineIndex is the line number (0-based).
 */
function parseHeadings(article: string): Array<{ text: string; lineIndex: number }> {
  const lines = article.split(/\r?\n/);
  const headings: Array<{ text: string; lineIndex: number }> = [];
  for (let i = 0; i < lines.length; i++) {
    const m = lines[i].match(/^#{1,6}\s+(.+)$/);
    if (m) headings.push({ text: m[1].trim(), lineIndex: i });
  }
  return headings;
}

/**
 * Find the line index of the heading matching the anchor section.
 * Strategy: exact match -> slug match -> startsWith (slug).
 * Returns -1 if not found.
 */
function resolveHeadingLineIndex(headings: Array<{ text: string; lineIndex: number }>, section: string): number {
  const targetSlug = toSlug(section);
  // 1. Exact match
  let found = headings.find((h) => h.text === section);
  if (found) return found.lineIndex;
  // 2. Slug match
  found = headings.find((h) => toSlug(h.text) === targetSlug);
  if (found) return found.lineIndex;
  // 3. StartsWith slug match
  found = headings.find((h) => toSlug(h.text).startsWith(targetSlug));
  if (found) return found.lineIndex;
  return -1;
}

export interface InsertResult {
  html: string;
  errors: Array<{ code: string; message: string }>;
}

/**
 * Inserts image/widget HTML blocks anchored to headings.
 * This is the v2 placement engine - replaces paragraph-index based insertPlacements.
 */
export function insertAnchoredPlacements(
  article: string,
  brief: ContentBrief,
  placements: AnchoredPlacement[],
  siteKey: string
): InsertResult {
  const errors: Array<{ code: string; message: string }> = [];

  if (placements.length === 0) {
    return { html: article, errors };
  }

  const lines = article.split(/\r?\n/);
  const headings = parseHeadings(article);

  // Collect insertion instructions (line index, html to insert after)
  type Insertion = { afterLine: number; html: string };
  const insertions: Insertion[] = [];

  const totalWidgets = placements.filter((p) => p.type === "widget").length;
  let widgetCount = 0;

  for (const placement of placements) {
    let insertAfterLine: number;

    if (placement.anchor.kind === "after-intro") {
      // Insert after the first blank line following the first heading
      const firstHeading = headings[0];
      if (!firstHeading) {
        errors.push({ code: "anchor_unresolved", message: `No heading found for after-intro placement of "${placement.productId}"` });
        continue;
      }
      // Find next non-empty content after the heading
      insertAfterLine = firstHeading.lineIndex + 1;
      while (insertAfterLine < lines.length && lines[insertAfterLine].trim() === "") insertAfterLine++;

    } else if (placement.anchor.kind === "end-of-section") {
      const headingLineIdx = resolveHeadingLineIndex(headings, placement.anchor.section);
      if (headingLineIdx === -1) {
        errors.push({ code: "anchor_unresolved", message: `Heading not found for end-of-section: "${placement.anchor.section}" (product: ${placement.productId})` });
        continue;
      }
      // Find next heading after this one, insert before it
      const nextHeadingIdx = headings.findIndex((h) => h.lineIndex > headingLineIdx);
      insertAfterLine = nextHeadingIdx === -1
        ? lines.length - 1
        : headings[nextHeadingIdx].lineIndex - 1;

    } else if (placement.anchor.kind === "after-heading") {
      const headingLineIdx = resolveHeadingLineIndex(headings, placement.anchor.section);
      if (headingLineIdx === -1) {
        errors.push({ code: "anchor_unresolved", message: `Heading not found for after-heading: "${placement.anchor.section}" (product: ${placement.productId})` });
        continue;
      }
      insertAfterLine = headingLineIdx;

    } else if (placement.anchor.kind === "before-heading") {
      const headingLineIdx = resolveHeadingLineIndex(headings, placement.anchor.section);
      if (headingLineIdx === -1) {
        errors.push({ code: "anchor_unresolved", message: `Heading not found for before-heading: "${placement.anchor.section}" (product: ${placement.productId})` });
        continue;
      }
      insertAfterLine = Math.max(0, headingLineIdx - 1);

    } else {
      errors.push({ code: "anchor_unresolved", message: `Unknown anchor kind: "${(placement as AnchoredPlacement).anchor.kind}"` });
      continue;
    }

    let html: string;
    if (placement.type === "image") {
      html = renderImage(placement.productId, brief);
    } else {
      const variant: WidgetVariant = widgetCount % 2 === 0 ? "singleproduct" : "product";
      html = renderWidget(placement.productId, brief, siteKey, variant);
      widgetCount++;
    }

    if (html) insertions.push({ afterLine: insertAfterLine, html });
  }

  // Apply insertions in descending line order to preserve indices
  insertions.sort((a, b) => b.afterLine - a.afterLine);
  for (const { afterLine, html } of insertions) {
    lines.splice(afterLine + 1, 0, html);
  }

  return { html: lines.join("\n"), errors };
}


type WidgetVariant = "singleproduct" | "product";

/** Strip "pr_" prefix → numeric PriceRunner product ID */
function numericId(productId: string): string {
  return productId.replace(/^pr_/, "");
}

// ─── Image HTML ───────────────────────────────────────────────────────────────
function renderImage(
  productId: string,
  brief: ContentBrief
): string {
  const product = brief.products.find((p) => p.id === productId);
  const image = brief.images.find((img) => img.productId === productId);
  if (!product || !image?.url) return "";

  const brand = product.specs["brand"];
  const alt = brand ? `${product.name} - ${brand}` : product.name;

  return `\n\n<figure class="my-8">
  <img src="${image.url}" alt="${alt}" loading="lazy" class="mx-auto rounded-xl shadow-md border border-gray-100" />
  <figcaption class="text-center text-sm text-gray-500 mt-3 font-medium">${image.caption}</figcaption>
</figure>\n\n`;
}

// ─── Widget HTML ──────────────────────────────────────────────────────────────

/**
 * Renders the actual PriceRunner JS embed widget.
 *  - "singleproduct" → singleproduct.js  (lowest price, best for hero/SPR star product)
 *  - "product"       → product.js         (up to 3 offers, best for alternatives/roundups)
 * Falls back to a styled Tailwind card when partnerId is not configured.
 */
function renderWidget(
  productId: string,
  brief: ContentBrief,
  siteKey: string,
  variant: WidgetVariant
): string {
  const product = brief.products.find((p) => p.id === productId);
  if (!product) {
    console.warn(`[widget-inserter] Unknown product id: ${productId} - skipping`);
    return "";
  }

  const siteConfig = SITE_CONFIGS[siteKey];
  const partnerId = siteConfig?.pricerunnerPartnerId ?? "";
  const country = (siteConfig?.pricerunnerCountry ?? "DK").toLowerCase();

  if (!partnerId) {
    return renderTailwindFallback(product);
  }

  const pid = numericId(product.id);
  const encodedPartnerId = encodeURIComponent(partnerId);
  const widgetElemId = `pr-${variant}-widget-${randomUUID().slice(0, 8)}`;
  const scriptFile = variant === "product" ? "product.js" : "singleproduct.js";
  const extraParams = variant === "product"
    ? "&onlyInStock=true&offerOrigin=NATIONAL&offerLimit=3"
    : "";

  const sep = product.affiliateUrl.includes("?") ? "&" : "?";
  const productUrlWithPartner = `${product.affiliateUrl}${sep}partnerId=${encodedPartnerId}`;

  return `\n\n<div class="pr-widget-wrapper my-8">
<div id="${widgetElemId}" style="display: block; width: 100%"></div>
<script type="text/javascript" src="https://api.pricerunner.com/publisher-widgets/${country}/${scriptFile}?productId=${pid}&partnerId=${encodedPartnerId}&widgetId=${widgetElemId}${extraParams}" async></script>
<div style="display: inline-block; margin-top: 4px;">
  <a href="${productUrlWithPartner}" rel="sponsored nofollow">
    <p style="font: 14px 'Klarna Text', Helvetica, sans-serif; font-style: italic; color: #888; text-decoration: underline; margin: 0;">Annonce i samarbejde med <span style="font-weight: bold">PriceRunner</span></p>
  </a>
</div>
</div>\n\n`;
}

/** Fallback card used during local dev when PR_*_PARTNER_ID env var is not set */
function renderTailwindFallback(
  product: ContentBrief["products"][number]
): string {
  return `\n\n<div class="price-widget not-prose flex flex-col md:flex-row items-center justify-between p-6 bg-white border border-gray-200 rounded-2xl shadow-sm my-8 hover:shadow-md transition-shadow gap-6">
  <div class="flex flex-col">
    <span class="text-gray-400 text-xs uppercase tracking-widest font-bold mb-1">Prissammenligning</span>
    <h3 class="text-xl font-bold text-gray-900">${product.name}</h3>
    <div class="flex items-center gap-2 mt-2">
      <span class="text-sm font-medium text-gray-600">${product.retailer}</span>
      <span class="inline-flex items-center px-2.5 py-0.5 rounded-full text-xs font-medium bg-green-100 text-green-800">På lager</span>
    </div>
  </div>
  <div class="flex items-center gap-6">
    <div class="text-right">
      <div class="text-gray-500 text-sm mb-0.5">Fra</div>
      <div class="text-3xl font-black text-orange-600 tracking-tighter">${product.priceKr} kr.</div>
    </div>
    <a href="${product.affiliateUrl}" rel="sponsored" class="buy-btn inline-flex items-center gap-2 bg-orange-500 hover:bg-orange-600 text-white px-6 py-3 rounded-xl font-bold text-base transition-all shadow-md shadow-orange-100 whitespace-nowrap">
      Sammenlign priser &rarr;
    </a>
  </div>
</div>\n\n`;
}

// ─── Agent-directed placement engine ─────────────────────────────────────────
/**
 * Injects agent-specified image and widget HTML blocks into a Markdown article.
 * Placements are applied in descending after_paragraph order to preserve earlier indices.
 *
 * Widget variants alternate by article position:
 *   - 1st widget in article → "singleproduct" (lowest price, strong CTA)
 *   - 2nd widget in article → "product" (top 3 offers, good for alternatives)
 *   - 3rd widget → "singleproduct", etc.
 */
export function insertPlacements(
  article: string,
  brief: ContentBrief,
  placements: Placement[],
  siteKey: string
): string {
  const paragraphs = article.split(/\r?\n\s*\r?\n/);

  // Sort descending so earlier paragraph indices remain valid after each insertion
  const sorted = [...placements].sort((a, b) => b.after_paragraph - a.after_paragraph);

  // Pre-assign widget variants in article order (ascending paragraph).
  // We process descending, so widgetIdx starts at (total-1) and counts down.
  // Index 0 (first in article) → "singleproduct"; index 1 → "product"; alternating.
  const totalWidgets = placements.filter((p) => p.type === "widget").length;
  let widgetIdx = totalWidgets - 1;

  for (const placement of sorted) {
    let html: string;

    if (placement.type === "image") {
      html = renderImage(placement.productId, brief);
    } else {
      const variant: WidgetVariant = widgetIdx % 2 === 0 ? "singleproduct" : "product";
      html = renderWidget(placement.productId, brief, siteKey, variant);
      widgetIdx--;
    }

    if (!html) continue;

    const insertAt = Math.min(placement.after_paragraph, paragraphs.length);
    paragraphs.splice(insertAt, 0, html);
  }

  return paragraphs.join("\n\n");
}

// ─── Legacy sync export (kept for backward compat with test/route code) ───────
export function insertWidgets(article: string, brief: ContentBrief): string {
  return insertPlacements(article, brief, [], "techblog");
}
